오픈마켓 I [STEP 2] 그루트, 바드

bar-d opened the pull request:
안녕하세요 찰리! @kcharliek 🍀
바드@bar-d 와 그루트@Groot-94 입니다! 🧏‍♂️
Step 2 PR 보냅니다🙃
컬렉션 뷰를 처음 사용해봐서 아직 많이 어렵네요..!
저희가 컬렉션 뷰를 잘 활용했는지가 궁금하네요
이번 PR도 잘 부탁드리겠습니다...!

## 📌 배경
-  URL Session을 활용한 서버와의 통신을 통해 전달받은 JSON 데이터와 매핑할 모델 활용
-  Modern CollectionView를 이용해 List, Grid 방식으로 마켓의 물품을 표시하도록 구현
## 📌 작업 내용
|레이아웃 변경 화면|화면 스크롤|
|:---:|:---:|
| <img src = "https://i.imgur.com/M2Ibdfr.gif" width="300" height="600">| <img src = "https://i.imgur.com/ZEAimHs.gif" width="300" height="600"> |
#### UISegmentedControl
- UISegmentedControl을 통한 화면 전환
#### UICollectionView
- Mordern Collection View 방식의 List Configuration을 활용한 List layout 컬렉셔 뷰
- Mordern Collection View 방식의 Compositional Layout 활용한 Grid layout 컬렉션 뷰
#### UIActivityIndicatorView
- 로딩 상태를 나타내기 위한 비동기식 뷰 구현
## 📌 테스트 방법
- 앱을 실행하여 기능을 테스트
## 📌 리뷰 노트
#### Ambigious UILabel height in autoresizing UITableViewCell 경고창이 뜨는 문제
- StackView에 Label을 순서대로 넣은 상태에서 label들의 text만 정해주면 따로 label들의 constraints 설정없이 레이아웃이 정렬된다고 생각했습니다.
    - 하지만, 그림과 같은 경고창을 보여줬습니다.
    <img src = "https://i.imgur.com/Tm4zsMo.png" width="500" height="400">
    
- [해당 블로그](http://47.112.232.56/a/stackoverflow/en/629df61c4deb6d4bcd5064d5.html)를 보고 Content Hugging priority를 설정해 보았지만 해결이 되지 않았었습니다🥲
[UIStackView.Distribution.equalCentering](https://developer.apple.com/documentation/uikit/uistackview/distribution/equalcentering)

- 여러가지 방법을 사용해서 경고창을 없애려고 했지만 결국에 찾은 방법은 Stack View의 distribution을 변경해주는 방법이였습니다.
    - 변경 전
    <img src = "https://i.imgur.com/jPVepIj.png" width="700" height="200">  

    - 변경 내용
    <img src = "https://i.imgur.com/oQUPFvV.png" width="700" height="200">
    
[UIStackView.Distribution.equalCentering - 공식문서](https://developer.apple.com/documentation/uikit/uistackview/distribution/equalcentering)

#### UIView with rounded corners and drop shadow / Modern Collection View's grid layout 사용 시 cell에 테두리가 자동으로 설정된다고 생각했으나, 테두리가 나타나지 않음.
- grid view를 구현하면서 cell의 테두리가 나타나지 않아서 Modern Collection View's Grid layout에 대해 찾아봤지만 테두리에 대한 내용을 찾지 못했습니다.
- 결국 cell의 draw 함수를 override 해서 layout 프로퍼티를 이용해 테두리를 구현하였습니다.
- Modern Collection View's grid layout에서 cell에 테두리를 설정하는 방법이 저희가 사용한 방법이 아닌,` Modern Collection View의 메서드나 프로퍼티를 이용해서 그려주는 방법이 있는지 궁금합니다.`

[참고 사이트](https://stackoverflow.com/questions/4754392/uiview-with-rounded-corners-and-drop-shadow)

Groot-94 committed:
docs: REAMDE.md 수정

bar-d committed:
refactor: 클래스 내 final 키워드 추가

Groot-94 committed:
refactor: Main Storyboard 제거 및 OpenMarketViewController 연결

bar-d committed:
feat: 세그먼트 컨트롤 네비게이션 컨트롤러 내부에 구현

bar-d committed:
Merge pull request #6 from bar-d/feat-3-controller

Feat 3 controller

Groot-94 committed:
feat: ListCollectionView 구현

Groot-94 committed:
Merge pull request #7 from bar-d/feat-5-ListCollectionView

feat: ListCollectionView 구현

bar-d committed:
feat: GridView 구현
- GridCollecionViewCell 구현

Groot-94 committed:
feat: GridView 테두리 추가 및 View 수정

bar-d committed:
feat:  로딩 뷰 구현

Groot-94 committed:
Merge pull request #8 from bar-d/feat-6-GridCollectionView

Feat 6 grid collection view

bar-d committed:
Merge pull request #9 from bar-d/feat-7-LoadingView

feat:  로딩 뷰 구현

Groot-94 committed:
refactor: extension 분리, price에 number formatter 적용

bar-d committed:
refactor: extension을 활용한 함수 분리

Groot-94 committed:
refactor: ListCollectionView, GridCollectionView 분리

bar-d committed:
feat: final 클래스 및 접근제한자 설정
- 마크다운 문법으로 함수 및 프로퍼티 구분

Groot-94 committed:
refactor: magic literal, magic number를 name space로 정리

Groot-94 committed:
refactor: mark 추가 및 name space 추가

bar-d committed:
Merge pull request #10 from bar-d/refactor-2-step2

Refactor 2 step2

bar-d committed:
refactor: 불필요한 구문 제거

bar-d committed:
Merge pull request #11 from bar-d/refactor-2-step2

refactor: 불필요한 구문 제거

bar-d committed:
refactor: 함수 순서 변경

bar-d committed:
Merge pull request #12 from bar-d/refactor-2-step2

refactor: 함수 순서 변경

kcharliek commented:
## 답변

1. size가 ambiguous 하다고 말하면, 반대로 explicit 하게 지정해주면 해결됩니다. 가장 쉬운 방법은 constraint를 주는 것이지요. height/ y position / top / bottom 이 모호하다고 하면 height 에 constraint를 걸어주면 되고, width / x / leading / trailing 이 모호하다고 하면 width에 constraint를 걸어주면 대부분 잡힙니다. 

2. 테두리 (border) 와 관련해서. Modern CollectionView에서 관련 설정을 해본 경험이 없어서 정확한 답변은 어렵습니다.
다만 draw를 이용한 해결은 가급적 피해야하는 방향입니다. 왜냐면 굉장히 강력한 무기 이기 때문에, 잘못 쓰면 성능 등의 사이드이펙트가 발생하기 쉽기 때문입니다.
우선 Border를 그리겠다고 마음 먹은 상태에서 답변을 드릴게요. 
(1) view를 이용해서 line을 만들어 주는 방법이 있습니다. 단순무식하지만 꽤 효율적일 때가 많습니다.
(2) layer.borderWidth를 이용하는 방법이 있습니다. 꽤 명확한 지정 방식이지요.
어지간 하면 1, 2,번에서 마무리가 되실거고요. 여기서 부족한 경우라면 최종의 최종 방법으로 draw를 이용해주세요 .

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,124 @@
+final class OpenMarketViewController: UIViewController {
+    // MARK: - properties
+    
+    var loadingView : UIView?
```
property 전부 private으로 감춰주세요 . 특히 UI property는 외부로 노출하지 않는 것이 좋습니다.
property가 오픈된 상황은 로직을 '통제' 하지 못하게 합니다.

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,124 @@
+        return segmentedControl
+    }()
+    
+    private var shouldHideListView: Bool? {
```
didSet 을 사용한 방식은, 사용자가 didSet 동작을 예상하지 못할 수 있다는 우려가 있습니다.

우리는 직접 만들었으니, shouldHideListView = someValue를 하면 뭔가 더 처리가 된다는 인식이 있지만,
이 코드를 처음 쓰는 사람은 값 설정 만으로 무언가 처리될 것이라는 기대가 없을 수 있어요.

저는 이런 이유로 가급적 didSet을 피합니다. 저와 함께하는 리뷰에서는 didSet 을 한번 피해보겠습니다. 

별도 method로 분리해주세요.

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,124 @@
+            (result: Result<ProductsDetailList, Error>) in
+            switch result {
+            case .success(let success):
+                for number in 0..<Product.itemPerPage.number {
```
for - in 구문은 피하겠습니다.

for-each 구문으로 수정해주세요 ,.

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,124 @@
+    }
+    
+    private func removeSpinner() {
+        DispatchQueue.main.async {
```
DispatchQueue.main.async 처리는 removeSpinner가 할 것이 아니라,
main thread를 보장하지 못하는 API Block에서 걸어주는 것이 좋을 것 같습니다.

main thread가 이미 보장되는 다른 곳에서도 removeSpinner 를 쓸 수 있으니까요 ,.

kcharliek commented on OpenMarket/OpenMarket/Model/Data/ProductsRequest.swift:
```diff
@@ -0,0 +1,24 @@
+        URLHost.openMarket.url + path.value
+    }
+    var headers: [String : String]?
+    var query: [URLQueryItem]? {
```
URLQueryItem을 직접 넘기는 방식이 아닌,
Dictionary로 넘기는 방식으로 수정해보겠습니다.
``` swift
var query: [String: String] {
    Product.page.text:  "\(Product.page.number)",
   Product.itemPerPage.text: "\(Product.itemPerPage.number)"
}
```

kcharliek commented on OpenMarket/OpenMarket/Model/Extension/NSMutableAttributedString+extension.swift:
```diff
@@ -0,0 +1,31 @@
+import UIKit
+
+extension NSMutableAttributedString {
+    func makePriceText(string: String) -> NSMutableAttributedString {
```
파일 이름은 'NSMutableAttributedString+extension' 입니다.
그러기에 이 곳에는 NSMutableAttributedString 을 이용한 'Util' 이 존재하는 것이 자연스러울 것입니다.

그런데 구현된 메서드를 보면 Price와 관련된 '비즈니스로직' 이 구현되어있습니다.
위 메서드는 OpenMarket에서만 쓸 수 있는 method로, Util이라고 보기는 어렵습니다.

이러한 구현이 필요하다면, 사용되는 위치에서 private extension으로 구현해주세요 .

kcharliek commented on OpenMarket/OpenMarket/Model/Extension/ProductDetail+extension.swift:
```diff
@@ -0,0 +1,59 @@
+
+import UIKit
+
+extension ProductDetail {
```
파일 분리가 되면, 파일 길이가 짧아진다는 장점은 있습니다만.
ProductDetai 정의부를 봤을 때, setPriceText 와 같은 method가 구현되어있다는 사실을 알기 어렵다는 단점도 있습니다.

kcharliek commented on OpenMarket/OpenMarket/View/GridCollecntionView.swift:
```diff
@@ -0,0 +1,55 @@
+final class GridCollecntionView: UICollectionView {
+    // MARK: - properties
+    
+    var gridViewDataSource: UICollectionViewDiffableDataSource<Section, ProductDetail>? = nil
```
private 전부 챙겨주세요.

별거 아닌 것 같지만, 협업할 때에는 굉장히 중요합니다.

kcharliek commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,94 @@
+final class GridCollectionViewCell: UICollectionViewCell {
+    // MARK: - properties
+    
+    var productImage: UIImageView = {
```
private 처리 전부 해주세요.

그렇게되면 cell 에 데이터를 주입하는 코드를 지금과는 다르게 해야할 것 입니다.

한번 해보세요 .

kcharliek commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,94 @@
+    // MARK: - functions
+    
+    override func draw(_ rect: CGRect) {
+        self.layer.cornerRadius = 10
```
이러한 구현은 draw로 할 필요가 없습니다. 

init쪽에서 구현해주세요.

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,124 @@
+extension OpenMarketViewController {
+    private func showSpinner(on view : UIView) {
+        let spinnerView = UIView.init(frame: view.bounds)
+        spinnerView.backgroundColor = UIColor.init(red: 0.5, green: 0.5, blue: 0.5, alpha: 0.5)
```
상수 전부 분리하겠습니다.

``` swift
private enum Design {
    static let spinnerBackgroundColor =UIColor.init(red: 0.5, green: 0.5, blue: 0.5, alpha: 0.5)
}

spinnerView.backgroundColor = Design.spinnerBackgroundColor
```

상수(리터럴) 은 그 자체로 의미가 없기 때문에, 라벨링이 필요하기도 하고요.
상수를 모아서 관리했을 때, 유지보수가 쉽다는 장점도 있습니다.

kcharliek commented on OpenMarket/OpenMarket/View/GridCollecntionView.swift:
```diff
@@ -0,0 +1,55 @@
+        self.configureDataSource()
+    }
+    
+    required init?(coder: NSCoder) {
```
가급적 init?(coder 를 구현 하셨으면 좋겠습니다.

언제 누가 InterfaceBuilder를 사용할지 모르니까요 . 
기능을 제공하고 쓰지 못하게 하는 것과, 애초에 구현을 안해서 기능 자체를 제공하지 않는 것은 또 다른 느낌입니다.

정 쓰기 싫게 하고싶으시다면,
@available(*, available) 을 붙여주세요 .

Groot-94 committed:
refactor: 접근제한자 설정

Groot-94 committed:
Revert "refactor: 접근제한자 설정"

This reverts commit 1d0fb1ebee8a1e5ec98ba54e4bc55e070eae9a3f.

Groot-94 committed:
refactor: 접근제한자 설정

bar-d committed:
refactor:  shoulHideListView 프로퍼티 제거
- 프로퍼티를 제거해주고 segment.selectedSegmentIndex로 switch case 구문 활용

Groot-94 committed:
refactor: for-in 구문 for-Each 구문으로 수정

bar-d committed:
refactor: showSpinner, removeSpinner DispatchQueue를 APIBlock 함수에서 실행

Groot-94 committed:
refactor: APIRequest의 query 타입 수정

bar-d committed:
refactor: ProductsDetailList에 필요한 extenstion들 수정

Groot-94 committed:
refactor: GridCollectionViewCell의 initailizer 수정

Groot-94 committed:
refactor: List, GridCollecntionView의 required init? 구현

bar-d committed:
style: 매직 넘버 제거

Groot-94 committed:
refactor: ProductDetail extension 위치 수정 및 setPriceText 메서드 수정

- bargain price와 price가 같으면 price만 표시

bar-d committed:
refactor: magic number 리팩토링

Groot-94 committed:
Merge pull request #13 from bar-d/refactor-2-step2

Refactor 2 step2

Groot-94 committed:
feat: refresh controll 구현

bar-d committed:
Merge pull request #14 from bar-d/feat-8-RefreshControl

feat: refresh controll 구현

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,156 @@
+    private func fetchData() {
+        let productsRequest = ProductsRequest()
+        let myURLSession = MyURLSession()
+        myURLSession.dataTask(with: productsRequest) {
```
weak capture가 되어야합니다.

Groot-94 replied:
수정했습니다.

[7606dd5](https://github.com/yagom-academy/ios-open-market/pull/184/commits/7606dd5997f6845a3977e66c2bf609fa53e7a230)

kcharliek commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,156 @@
+    // MARK: - @objc functions
+    
+    @objc func segmentButtonDidTap(sender: UISegmentedControl) {
+        switch sender.selectedSegmentIndex {
```
enum SectionType {
   case grid
   case list
}

guard let type = SectionType(rawValue: sender.selectedSegmentIndex) else { return }

Groot-94 replied:
Section enum의 init?을 사용해서 케이스로 구분하도록 수정했습니다.

[e480f9a](https://github.com/yagom-academy/ios-open-market/pull/184/commits/e480f9a99ede9edf8c6d5b0e25982bf56932b7ad)

kcharliek commented on OpenMarket/OpenMarket/Model/JSONModel/ProductDetail.swift:
```diff
@@ -32,3 +34,77 @@ struct ProductDetail: Codable {
+        }
+    }
+    
+    func setThumbnailImage() -> UIImage {
```
get/fetch/make / create / generate (o)
set (x)

kcharliek replied:
Data(contentsOf: url) 를 비동기 코드로 바꿔주세요 .

kcharliek replied:
iOS Swift UIImage From URL Async

Groot-94 replied:
저희가 만든 APIRequest를 수정해서 request를 통해 서버에서 image를 받아오는 방법으로 수정했습니다.

[1ad86e9](https://github.com/yagom-academy/ios-open-market/pull/184/commits/1ad86e9351516ac05aa087c8f510b07b4829bb40)

kcharliek commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+    
+    func setViewItems(_ item: ProductDetail) {
+        productImage.image = item.setThumbnailImage()
```
setThumbnailImage 를 비동기로 변경했을 떄 재사용 이슈가 발생할 수 있습니다.

prepareForReuse로 이를 해결해주세요,

Groot-94 replied:
gridViewCell의 경우 prepareForReuse를 사용해서 각 컨텐츠에 nil을 할당해주는 방법으로 해결했습니다. 하지만, [prepareForReuse공식문서](https://developer.apple.com/documentation/uikit/uitableviewcell/1623223-prepareforreuse)에 
` To avoid potential performance issues, you should only reset attributes of the cell that are not related to content, for example, alpha, editing, and selection state.` 와 같은 문구가 있는걸 확인했습니다. 
저희가 nil을 할당해주는 동작도 컨텐츠를 조작한다고 생각이 들었는데, 단순히 nil을 할당해주는 동작은 성능이슈가 없을지 궁금합니다.

또한, list view 같은 경우는 UIListContentConfiguration의 기본 아이템을 사용해서 셀을 구성해주고 있습니다. 이 경우 기본 cell을 사용하기 때문에 prepareForReuse를 호출하지 않아도 재사용 이슈가 없을지, 그게 아니라 Custom Cell을 만들어서 재사용 이슈를 없애줘야 하는지 궁금합니다.

[de58a11](https://github.com/yagom-academy/ios-open-market/pull/184/commits/de58a11065cbf83226b628d1337a61cd331f8029)

kcharliek replied:
이 내용에 대해서 저희 팀에서 연구해본적이 있는데요. 결론은 어지간한 reset으로는 '큰 영향 없다' 는 것 이었습니다.
애플의 말은 prepareForReuse 시점에 UI 업데이트를 해서 main thread 부하를 주지 말라는 것인데요 . 만일 부하가 누적된다면 스크롤이 끊길 수 있기 때문입니다.

그런데 우리의 상황인 image = nil 정도로는 main thread에 부하를 주지 못합니다. AutoLayout 업데이트와 같은 것을 하지 말라는 것으로 이해해주시면 좋겠습니다 .

기본 Cell과 관련해서는 상황 마다 다르겠습니다만. 대부분의 재사용 관련 문제가 발생하는 이미지의 경우, cell configure시점에 API 통신 (서버에서 이미지를 가져온다던지) 을 해야한다면, reset이 필요해요~

Groot-94 committed:
docs: README.md 수정

Groot-94 committed:
refactor: APIRequest 타입의 urlQueryItems 프로퍼티 수정

bar-d committed:
refactor: OpenMarketViewController extension 제거

Groot-94 committed:
refactor: OpenMarketViewController 리뷰 적용

- 메서드 네임 수정
- fetchData 메서드 로직 수정
- segmentButtonDidTap 메서드 로직 수정

bar-d committed:
refactor: 클로저 내부 weak 캡쳐 추가

Groot-94 committed:
refactor: setting 관련 메서드 네이밍 수정

- ProductDetail
- GridCollecntionView
- GridCollectionViewCell
- ListCollectionView

bar-d committed:
refactor: 이미지 캐싱을 위한 리팩토링

Groot-94 committed:
fix: 셀에 image가 나오지 않는 오류 수정

bar-d committed:
Merge pull request #15 from bar-d/refactor-3-step2review

Refactor 3 step2review

kcharliek commented on OpenMarket/OpenMarket/Model/Data/ImageCacheManager.swift:
```diff
@@ -0,0 +1,14 @@
+import UIKit
+
+class ImageCacheManager {
+    static let shared = NSCache<NSString, UIImage>()
```
Singleton pattern의 기본 패턴은

``` swift

class A {
   static let shared = .init()
}

```

입니다.

만일 위와 같이 NSCache 객체를 들고있고 싶다면,

``` swift

class A {
   static let shared = .init()

   private let cache = NSCache<NSString, UIImage>()

   func object(for key: String) {}
}

```

와 같이 NSCache 를 숨기고,  메서드를 열어서 기능을 제공하는 식이죠 .

kcharliek commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,131 @@
+final class GridCollectionViewCell: UICollectionViewCell {
+    // MARK: - properties
+    
+    private var productImage: UIImageView = {
```
var -> let 으로 수정 되면 좋겠습니다.

kcharliek commented on OpenMarket/OpenMarket/View/GridCollectionViewCell.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+    
+    func setViewItems(_ item: ProductDetail) {
+        if let image = ImageCacheManager.shared.object(forKey: NSString(string: item.thumbnail)) {
```
Repository 와 같은 Layer가 있으면 좋겠네요. 

OpenMarketRepository 정도를 두어, 이미지 캐싱 로직을 내부로 숨기는거죠~ 

OpenMarketRequest 를 호출할 때 마다 반복되는 ImageCaching 로직을 한번만 구현해서 재사용 하도록 할 수도 있고요 .

kcharliek reviewed (approved):
잘 해주셨습니다 . 

추가 코멘트 드린 부분은 다음 스텝에서 진행해주세요:)

고생 많으셨습니다~!
